core/src/client/async_client/mod.rs

```diff
@@ -395,172 +604,59 @@ async fn background_task<S, R>(
 		let res = receiver.receive().await;
 		Some((res, receiver))
 	});
-
 	futures_util::pin_mut!(backend_event);
 
-	loop {
-		let next_frontend = frontend.next();
-		let next_backend = backend_event.next();
-		futures_util::pin_mut!(next_frontend, next_backend);
-
-		match futures_util::future::select(next_frontend, next_backend).await {
-			// User dropped the sender side of the channel.
-			// There is nothing to do just terminate.
-			Either::Left((None, _)) => {
-				tracing::trace!("[backend]: frontend dropped; terminate client");
-				break;
-			}
-
-			Either::Left((Some(FrontToBack::Batch(batch)), _)) => {
-				tracing::trace!("[backend]: client prepares to send batch request: {:?}", batch.raw);
-				// NOTE(niklasad1): annoying allocation.
-				if let Err(send_back) = manager.insert_pending_batch(batch.ids.clone(), batch.send_back) {
-					tracing::warn!("[backend]: batch request: {:?} already pending", batch.ids);
-					let _ = send_back.send(Err(Error::InvalidRequestId));
-					continue;
-				}
-
-				if let Err(e) = sender.send(batch.raw).await {
-					tracing::warn!("[backend]: client batch request failed: {:?}", e);
-					manager.complete_pending_batch(batch.ids);
-				}
-			}
-			// User called `notification` on the front-end
-			Either::Left((Some(FrontToBack::Notification(notif)), _)) => {
-				tracing::trace!("[backend]: client prepares to send notification: {:?}", notif);
-				if let Err(e) = sender.send(notif).await {
-					tracing::warn!("[backend]: client notif failed: {:?}", e);
-				}
-			}
+	// Place frontend and backend messages into their own select.
+	// This implies that either messages are received (both front or backend),
+	// or submit ping timer expires (if provided).
```

Comment:
```suggestion
	// or the submitted ping timer expires (if provided).
```

I was confused at first when reading this in my head but this is how it sounded when i read it out loud.